### ISSUE TITLE
Change webserver port

### ISSUE THREAD

**Issue by fawzibr** [Mailu/Mailu#64]
When I set FRONTEND and/or WEBMAIL to none, it gives an error:

Pulling repository docker.io/freeposte/none
ERROR: Tag stable not found in repository docker.io/freeposte/none

So it seems i cant disable the webserver and it binds to port 80. How about adding an option to change the webserver port?

**Comment by kaiyou** [comment]
Indeed, the automated build of the ``none`` image was failing on Docker Hub. This should be fixed now. There is still an issue about Docker binding ports 80 and 443 even if not used by Freeposte.

Any idea about making this conditional is welcome.

**Comment by joeknock90** [comment]
Would using expose in the docker-compose.yml work to get it to not bind on 80 and 443. As long as the ports are exposed in the container you should be able to use a separate nginx (like jwilder's nginx w/ docker-gen)

I don't know if I'm getting the wrong idea here.

**Comment by curry684** [comment]
Isn't it easier to provide an installer script that generates the docker-compose.yml file following some basic questions? Saves all the hassle with the none container which is really an ugly hack as well.

**Comment by kaiyou** [comment]
@joeknock90 modifying the ``docker-compose.yml`` will work of course. For instance you could expose port 80 from the Webmail and admin interface on ``localhost`` and point your proxy there.

@curry684 I would have loved to find a proprer solution that does not involve generating a custom ``docker-compose.yml``. Problem with this approach is: if users want to switch from one Webmail to another or disable the reverse proxy, they will have to manually edit the ``docker-compose.yml``. It might be the only acceptable solution on the long run however.

**Comment by curry684** [comment]
That could be circumvented by storing the selection in an `options.conf` which you could then make editable in the admin. If you mount the Docker socket you could then even update automatically.

**Comment by fawzibr** [comment]
Wouldn't just making the http ports configurable as well solve this problem? Add 2 more variables to the .env file (something like HTTP_PORT=<value>,HTTPS_PORT=<value>) and using it in the docker-compose.yml? That way even if 'none' is used, just use some unused ports that wont conflict with anything also you wont need to generate the docker-compose.yml file.

**Comment by aminb** [comment]
@kaiyou Any updates on this? It'd be great if I the server I run freeposte on, I could also use for hosting my static blog (via nginx).

**Comment by kaiyou** [comment]
@aminb We have not made any decision yet. You can easily modify the ``front`` entry in your ``docker-compose.yml`` to disable the port forward for HTTP and HTTPS in the meanwhile.

**Comment by curry684** [comment]
Thinking out loud, possibly kicking up some controversies: the underlying issue here is that the project is monolithic, and thus violating the Docker base principle of *"batteries included but replaceable"*.

Right now I cannot change *any* part of the FreePoste setup without breaking my options of updating via Git easily. Most obviously the `.env` file should be distributed as `.env.dist` instead so every user can build their own `.env` from there and have it survive updates. At the more detailed level it's now non-trivial to plug in another webmail system, or drop it as a whole as this issue describes, without running into similar issues.

The solution is to split the project out into its separate parts. Make a separate organization on Github and Docker Hub, change the name while at it (FreePoste makes it sound like a bad ripoff of "the real thing" which it isn't), move all the separate images to separate repositories, enable automated public builds, and provide a `docker-compose.yml.dist` in a master repo that binds it all together in a default way. Every user can then for himself decide to either use the default setup, which is still a full featured mailserver in 5 minutes, or replace/remove any single part of it without sacrificing updateability. You could even make it so that the Nginx configs are expandable with smart volume mounts so the end user can add custom vhosts for their static blogs.

**Comment by dzervas** [comment]
I like this idea very much, but postfix and dovecot are the only players around (you definetly not want to dive into exim/sendmail/qmail/etc.). HTTP server & webmail are interchangable enough.

What I would like to be more interchangable is the filtering (spamassassing & clamav). ClamAV is kinda pricy in terms of RAM (as shown in #79 ) so I'd like to have an option to turn it off and spamassassing can be replaced by [rspamd](https://www.rspamd.com/) or turned off.

**Comment by curry684** [comment]
It doesn't matter that Postfix and Dovecot are the only manageable containers according to the maintainers. It matters that **all** batteries should be replaceable. If someone wants to plug in Exim instead of Postfix - all glory to them  😄

**Comment by dzervas** [comment]
I get what you sayin' but hey - lets focus on development that matters first
BTW: I'd really like to meet the guy that will play with sendmail on docker XD

**Comment by kaiyou** [comment]
Thanks @curry684 for your opinion and advice. I do agree that all batteries should be replaceable, which is one of the initial goals for Freeposte: to be both easy to setup and hacker-friendly.

Regarding the options:
- shipping ``docker-compose.yml`` and ``.env`` as dist files to be copied is definitely a great idea for people building from source on the server, will change this today,
- I do not see how spliting the repository will help, except in making things more complicated for tracking issues, I started Freeposte with multiple repositories and quickly switched to a single one, users are still free to modify the ``docker-compose.yml`` file and use any image they like or even add containers, splitting the repository will not change the architecture that I do not find too monolythic (multiple containers, multiple images, etc.)
- the name is purely a matter of opinion, I do like it, feel free to redistribute it under any name you think sounds better or suggest names in an effort to convince me
- about exposing some simple volumes to customize configuration, I agree and we already started wiht Postfix and Dovecot

Back to the original topic, @fawzibr I do not think that adding dozens of variables will help power users customize their email server in the end (there will always be complex cases that simple configuration does not cover). Configuration variables should target users who will use Freeposte in one of the default flavours (simple and secure by default, two or three lines of documentation to explain maybe how to replace a Webmail or enable an optional feature). More complex setups should be documented for power users and hackers to be able to properly customize their ``docker-compose.yml``.

**Comment by aminb** [comment]
+1 to @curry684 about name change.

Regarding ports, and with support for {Cal,Card}DAV (#68) around the corner, I suggest using [jwilder/nginx-proxy](/jwilder/nginx-proxy) and designate different subdomains for each instance:

- mail.DOMAIN (for email web frontend)
- dav.DOMAIN (for {Cal,Card}DAV web interface)
- ...

Here's a related [article](http://blog.florianlopes.io/host-multiple-websites-on-single-host-docker/).

This way, one can use the same server for multiple purposes (email, dav, static blog, ...).

**Comment by kaiyou** [comment]
On the subject of changing the port, I finally took some time and wrote a guide on hosting your own reverse proxy to setup some interesting configurations: https://github.com/Mailu/Mailu/wiki/Use-your-own-reverse-proxy

@aminb I disagree on the matter of using multiple hostnames. First, until letsencrypt is fully available, this would be a nightmare for users from a TLS point of view. Then, we are not hosting multiple services per se, but multiple endpoints to the same service: e-mail. Thus, it would make sense for 99% of our users to simply host their service as ``mail.domain.tld``.

In the guide I just wrote, I explain how a user can setup an additional reverse proxy on top of Mailu to serve other, distinct srevices, than e-mails.
